### PR TITLE
Add checks for common mistake of swapping component/custom resource options.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: sdk/nodejs/resource.ts [hunk 1]
```diff
@@ -299,10 +304,6 @@ export abstract class ProviderResource extends CustomResource {
      * @param opts A bag of options that control this provider's behavior.
      */
     constructor(pkg: string, name: string, props?: Inputs, opts: ResourceOptions = {}) {
-        if ((<any>opts).provider !== undefined) {
-            throw new ResourceError("Explicit providers may not be used with provider resources", opts.parent);
-        }
```
Comment (CyrusNajmabadi) [inline sdk/nodejs/resource.ts]: removed because it's redundant with base type check.

File: sdk/nodejs/resource.ts [hunk 2]
```diff
@@ -278,8 +278,13 @@ export abstract class CustomResource extends Resource {
...
      * @param opts A bag of options that control this resource's behavior.
      */
-    constructor(t: string, name: string, props?: Inputs, opts?: CustomResourceOptions) {
+    constructor(t: string, name: string, props?: Inputs, opts: CustomResourceOptions = {}) {
+        if ((<ComponentResourceOptions>opts).providers) {
+            throw new ResourceError("Do not supply 'providers' option to a CustomResource. Did you mean 'provider' instead?", opts.parent);
+        }
+
         super(t, name, true, props, opts);
+        this.__pulumiCustomResource = true;
```
Comment (swgillespie) [inline sdk/nodejs/resource.ts]: Why'd you move this out of an initializer?

Comment (CyrusNajmabadi) [reply sdk/nodejs/resource.ts]: annoyance in TS.  if you have a member initializer, you are not allowed to have pre-super checks.

File: sdk/nodejs/resource.ts [hunk 3]
```diff
@@ -278,8 +278,13 @@ export abstract class CustomResource extends Resource {
      * @param props The arguments to use to populate the new resource.
      * @param opts A bag of options that control this resource's behavior.
      */
-    constructor(t: string, name: string, props?: Inputs, opts?: CustomResourceOptions) {
+    constructor(t: string, name: string, props?: Inputs, opts: CustomResourceOptions = {}) {
+        if ((<ComponentResourceOptions>opts).providers) {
+            throw new ResourceError("Do not supply 'providers' option to a CustomResource. Did you mean 'provider' instead?", opts.parent);
```
Comment (lukehoban) [inline sdk/nodejs/resource.ts]: `opts.parent` will almost always be undefined, right?  I guess we aren’t allowed to use `this` here which would be the correct place to attach the error?

**Pull request merged**